src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/LdapDirectoryDAO.scala

```diff
@@ -42,58 +49,65 @@ class LdapDirectoryDAO(protected val ldapConnectionPool: LDAPConnectionPool, pro
-  }
+  private def unmarshalGroup(results: Entry): Either[String, BasicWorkbenchGroup] =
+    for {
+      cn <- getAttribute(results, Attr.cn).toRight(s"${Attr.cn} attribute missing: ${results.getDN}")
+      email <- getAttribute(results, Attr.email).toRight(s"${Attr.email} attribute missing: ${results.getDN}")
+      memberDns = getAttributes(results, Attr.uniqueMember)
+    } yield BasicWorkbenchGroup(WorkbenchGroupName(cn), memberDns.map(dnToSubject), WorkbenchEmail(email))
 
-  override def loadGroups(groupNames: Set[WorkbenchGroupName]): Future[Seq[BasicWorkbenchGroup]] = Future {
+  override def loadGroups(groupNames: Set[WorkbenchGroupName]): IO[Stream[BasicWorkbenchGroup]] = {
     val filters = groupNames.grouped(batchSize).map(batch => Filter.createORFilter(batch.map(g => Filter.createEqualityFilter(Attr.cn, g.value)).asJava)).toSeq
-    ldapSearchStream(groupsOu, SearchScope.SUB, filters:_*)(unmarshalGroup)
+
+    for {
+      streamOfEither <- cs.evalOn(ecForLdapBlockingIO)(IO(ldapSearchStream(groupsOu, SearchScope.SUB, filters: _*)(unmarshalGroup)))
+      res <- streamOfEither.parSequence.fold(err => IO.raiseError(new WorkbenchException(err)), r => IO.pure(r))
+    } yield res
   }
 
-  override def loadGroupEmail(groupName: WorkbenchGroupName): Future[Option[WorkbenchEmail]] = Future {
-    Option(ldapConnectionPool.getEntry(groupDn(groupName), Attr.email)) map { results =>
-      WorkbenchEmail(getAttribute(results, Attr.email).getOrElse(throw new WorkbenchException(s"${Attr.email} attribute missing: ${results.getDN}")))
-    }
+  override def loadGroupEmail(groupName: WorkbenchGroupName): IO[Option[WorkbenchEmail]] = {
+    for{
+      searchResult <- cs.evalOn(ecForLdapBlockingIO)(IO(ldapConnectionPool.getEntry(groupDn(groupName), Attr.email)))
+      res <- Option(searchResult).traverse{
+        e =>
+          getAttribute(e, Attr.email) match {
+            case Some(email) => IO.pure(WorkbenchEmail(email))
+            case None => IO.raiseError(new WorkbenchException(s"${Attr.email} attribute missing: ${e.getDN}"))
+          }
+      }
+    } yield res
   }
 
-  override def batchLoadGroupEmail(groupNames: Set[WorkbenchGroupName]): Future[Seq[(WorkbenchGroupName, WorkbenchEmail)]] = loadGroups(groupNames).map(_.map(g => g.id -> g.email))
+  override def batchLoadGroupEmail(groupNames: Set[WorkbenchGroupName]): IO[Stream[(WorkbenchGroupName, WorkbenchEmail)]] =
+    loadGroups(groupNames).map(_.map(g => g.id -> g.email))
 
-  override def deleteGroup(groupName: WorkbenchGroupName): Future[Unit] = {
+  override def deleteGroup(groupName: WorkbenchGroupName): Future[Unit] =
     listAncestorGroups(groupName).map { ancestors =>
       if (ancestors.nonEmpty) {
-        throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.Conflict, s"group ${groupName.value} cannot be deleted because it is a member of at least 1 other group"))
+        throw new WorkbenchExceptionWithErrorReport(
```

Comment:
did you mean to get rid of this throw?

Reply:
oh, I think this is one of those untouched but accidentally scalafmtted methods...I'll refactor this in a future PR